diff --git a/android/cmake/main.cmake b/android/cmake/main.cmake
index 2fa3ea0c3d7d9771729ad55de63185fdb90a3520..ff72a20c4fcb36e737e2fb2840cbf6a51a2e15d2 100644
--- a/android/cmake/main.cmake
+++ b/android/cmake/main.cmake
@@ -71,22 +71,31 @@ if (REACT_NATIVE_WORKLETS_DIR)
 if (REACT_NATIVE_WORKLETS_DIR)
-  add_library(worklets SHARED IMPORTED)
-
-  if (${CMAKE_BUILD_TYPE} MATCHES "Debug")
-    set(BUILD_TYPE "debug")
-  else ()
-    set(BUILD_TYPE "release")
-  endif ()
-
-  set_target_properties(
-    worklets
-    PROPERTIES
-    IMPORTED_LOCATION
-    "${REACT_NATIVE_WORKLETS_DIR}/android/build/intermediates/cmake/${BUILD_TYPE}/obj/${ANDROID_ABI}/libworklets.so"
-  )
-
-  target_link_libraries(
-    expo-modules-core
-    PRIVATE
-    worklets
-  )
+  find_package(react-native-worklets CONFIG QUIET)
+  if (react-native-worklets_FOUND)
+    target_link_libraries(
+      expo-modules-core
+      PRIVATE
+      react-native-worklets::worklets
+    )
+  else ()
+    add_library(worklets SHARED IMPORTED)
+
+    if (${CMAKE_BUILD_TYPE} MATCHES "Debug")
+      set(BUILD_TYPE "debug")
+    else ()
+      set(BUILD_TYPE "release")
+    endif ()
+
+    set_target_properties(
+      worklets
+      PROPERTIES
+      IMPORTED_LOCATION
+      "${REACT_NATIVE_WORKLETS_DIR}/android/build/intermediates/cmake/${BUILD_TYPE}/obj/${ANDROID_ABI}/libworklets.so"
+    )
+
+    target_link_libraries(
+      expo-modules-core
+      PRIVATE
+      worklets
+    )
+  endif ()
 endif ()
